### DISCUSSION TITLE
fillOutsideCircle() new function

### DISCUSSION THREAD

**Discussion by griffinmt** [Bodmer/TFT_eSPI#2724] [New processor/display support or capability request]
I am looking for a way to clear all pixels that are outside the area of a circle, sort of the inverse of fillSmoothCircle(). This would eliminate all data that was written that went past the edge of the circle. I tried making my own andd added it to TFT_eSPI.cpp but it couldnt find it when I compiled my application.
I would like it to work for both screen circles and sprite circles.
Can you help me with this?

Martyn

**Comment by Bodmer** [comment]
You could use fthe fillArc function with a 0 to 360 degree range (full circle). The inside radius being the circle diameter you do not want erased, and the outer radius being sufficient to cover the remainder of the screen/sprite.

The bad news with this approach is that the draw speed may not be enough for your needs as maths is done on every pixel. For typical small low pixel count arcs this is not much of a problem.  You could calculate the required outer radius from the size of the screen/sprite and the position of the centre of the required circle.

**Reply by griffinmt** [reply]
Timing should not be a problem as it only needs it done once. But I cannot find fillArc() defined in TFT_eSPI.cpp. Where would I find it so I can determine the correct parameters.
TIA

**Reply by griffinmt** [reply]
I found an example you posted online. And it almost works. but were the triangles extend past the edge of the screen, they wrap over to the othe edge. And how would I do this inside a sprite?

**Reply by griffinmt** [reply]
I added the following to the fillArc() routine just prior to the fillTriangle() statements:
    if (x0>480) x0=480;
    if (x0<0) x0=0;
    if (y0>320) y0=320;
    if (y0<0) y0=0;
    if (x2>480) x2=480;
    if (x2<0) x2=0;
    if (y2>320) y2=320;
    if (y2<0) y2=0;
    if (x3>480) x3=480;
    if (x3<0) x3=0;
    if (y3>320) y3=320;
    if (y3<0) y3=0;

Works!

**Comment by griffinmt** [comment]
There is a noticable delay when creating the sprite and I would like to eliminate that. I was thinging that I would create a backup sprite that is complete after the fillArc() function completes. Then I would create a new sprite each loop, but start by using the pushToSprite function to copy the backup into the new sprite.
That being said, it does not seem to let me display the result. Do you have any simple examples that would show how to accomplish this.

facebck.pushToSprite(&face,0,0);

TIA
Martyn.

**Comment by Bodmer** [comment]
The drawArc function is built into the latest library version. There is an example included, see here:
https://github.com/Bodmer/TFT_eSPI/blob/master/examples/Smooth%20Graphics/Draw_Arc/Draw_Arc.ino

Try using that built-in function as you may have picked up some early development code.

If you are still having problems then post a simple and complete example sketch that demonstrates the issue.

**Comment by griffinmt** [comment]
Works 'perfectly' for masking all the overrun. 
Is it possible to take the resulting sprite, and output it as a bmp or png image. I would either write it to SD, or save as an array in a .h file. Sorry for all the questions .

Would this be best done by readRect() to get the data, and pushRect() to redisplay it. If so, I could write the data to sd and read it later?

Better yet, how do I access the inram verion of a sprite and copy the 8 bit color values into a data array without using functions in the library
